File: agentwallet/packages/sdk-python/src/agentwallet/resources/acp.py
"""ACP (Agent Commerce Protocol) sub-resource."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..types import AcpJob, AcpMemo, ListResponse, ResourceOffering

if TYPE_CHECKING:
    from ..client import AgentWallet


class AcpResource:
    def __init__(self, client: AgentWallet):
        self._client = client

    # ── Jobs ──

    async def create_job(
        self,
        buyer_agent_id: str,
        seller_agent_id: str,
        title: str,
        description: str,
        price_usdc: float,
        service_id: str | None = None,
        evaluator_agent_id: str | None = None,
        requirements: dict | None = None,
        deliverables: dict | None = None,
        fund_transfer: bool = False,
        principal_amount_usdc: float | None = None,
    ) -> AcpJob:
        body: dict = {
            "buyer_agent_id": buyer_agent_id,
            "seller_agent_id": seller_agent_id,
            "title": title,
            "description": description,
            "price_usdc": price_usdc,
            "fund_transfer": fund_transfer,
        }
        if service_id:
            body["service_id"] = service_id
        if evaluator_agent_id:
            body["evaluator_agent_id"] = evaluator_agent_id
        if requirements:
            body["requirements"] = requirements
        if deliverables:
            body["deliverables"] = deliverables
        if principal_amount_usdc is not None:
            body["principal_amount_usdc"] = principal_amount_usdc
        data = await self._client.post("/acp/jobs", json=body)
        return AcpJob(**data)

    async def get_job(self, job_id: str) -> AcpJob:
        data = await self._client.get(f"/acp/jobs/{job_id}")
        return AcpJob(**data)

    async def list_jobs(
        self,
        agent_id: str | None = None,
        phase: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> ListResponse:
        params: dict = {"limit": limit, "offset": offset}
        if agent_id:
            params["agent_id"] = agent_id
        if phase:
            params["phase"] = phase
        data = await self._client.get("/acp/jobs", params=params)
        return ListResponse(
            data=[AcpJob(**j) for j in data["jobs"]],
            total=data["total"],
        )

    async def negotiate(
        self,
        job_id: str,
        seller_agent_id: str,
        agreed_terms: dict,
        agreed_price_usdc: float | None = None,
    ) -> AcpJob:
        body: dict = {"agreed_terms": agreed_terms}
        if agreed_price_usdc is not None:
            body["agreed_price_usdc"] = agreed_price_usdc
        data = await self._client._request(
            "POST", f"/acp/jobs/{job_id}/negotiate",
            json=body, params={"seller_agent_id": seller_agent_id},
        )
        return AcpJob(**data)

    async def fund(self, job_id: str, buyer_agent_id: str) -> AcpJob:
        data = await self._client._request(
            "POST", f"/acp/jobs/{job_id}/fund",
            params={"buyer_agent_id": buyer_agent_id},
        )
        return AcpJob(**data)

    async def deliver(
        self,
        job_id: str,
        seller_agent_id: str,
        result_data: dict,
        notes: str | None = None,
    ) -> AcpJob:
        body: dict = {"result_data": result_data}
        if notes:
            body["notes"] = notes
        data = await self._client._request(
            "POST", f"/acp/jobs/{job_id}/deliver",
            json=body, params={"seller_agent_id": seller_agent_id},
        )
        return AcpJob(**data)

    async def evaluate(
        self,
        job_id: str,
        evaluator_agent_id: str,
        approved: bool,
        evaluation_notes: str | None = None,
        rating: int | None = None,
    ) -> AcpJob:
        body: dict = {"approved": approved}
        if evaluation_notes:
            body["evaluation_notes"] = evaluation_notes
        if rating is not None:
            body["rating"] = rating
        data = await self._client._request(
            "POST", f"/acp/jobs/{job_id}/evaluate",
            json=body, params={"evaluator_agent_id": evaluator_agent_id},
        )
        return AcpJob(**data)

    # ── Memos ──

    async def send_memo(
        self,
        job_id: str,
        sender_agent_id: str,
        memo_type: str,
        content: dict,
        signature: str | None = None,
    ) -> AcpMemo:
        body: dict = {"memo_type": memo_type, "content": content}
        if signature:
            body["signature"] = signature
        data = await self._client._request(
            "POST", f"/acp/jobs/{job_id}/memos",
            json=body, params={"sender_agent_id": sender_agent_id},
        )
        return AcpMemo(**data)

    async def list_memos(self, job_id: str) -> ListResponse:
        data = await self._client.get(f"/acp/jobs/{job_id}/memos")
        return ListResponse(
            data=[AcpMemo(**m) for m in data["memos"]],
            total=data["total"],
        )

    # ── Resource Offerings ──

    async def create_offering(
        self,
        agent_id: str,
        name: str,
        description: str,
        endpoint_path: str,
        parameters: dict | None = None,
        response_schema: dict | None = None,
    ) -> ResourceOffering:
        data = await self._client.post("/acp/offerings", json={
            "agent_id": agent_id,
            "name": name,
            "description": description,
            "endpoint_path": endpoint_path,
            "parameters": parameters or {},
            "response_schema": response_schema or {},
        })
        return ResourceOffering(**data)

    async def list_offerings(
        self,
        agent_id: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> ListResponse:
        params: dict = {"limit": limit, "offset": offset}
        if agent_id:
            params["agent_id"] = agent_id
        data = await self._client.get("/acp/offerings", params=params)
        return ListResponse(
            data=[ResourceOffering(**o) for o in data["offerings"]],
            total=data["total"],
        )

File: agentwallet/packages/sdk-ts/src/resources/acp.ts
import type { AgentWallet } from '../client';
import type {
  AcpJob,
  AcpMemo,
  CreateAcpJobParams,
  CreateOfferingParams,
  ListResponse,
  ResourceOffering,
} from '../types';

export class AcpResource {
  constructor(private client: AgentWallet) {}

  // ── Jobs ──

  /** Create an ACP job */
  async createJob(params: CreateAcpJobParams): Promise<AcpJob> {
    const body: Record<string, unknown> = {
      buyer_agent_id: params.buyer_agent_id,
      seller_agent_id: params.seller_agent_id,
      title: params.title,
      description: params.description,
      price_usdc: params.price_usdc,
      fund_transfer: params.fund_transfer ?? false,
    };
    if (params.service_id) body.service_id = params.service_id;
    if (params.evaluator_agent_id) body.evaluator_agent_id = params.evaluator_agent_id;
    if (params.requirements) body.requirements = params.requirements;
    if (params.deliverables) body.deliverables = params.deliverables;
    if (params.principal_amount_usdc !== undefined) body.principal_amount_usdc = params.principal_amount_usdc;
    return this.client.post<AcpJob>('/acp/jobs', body);
  }

  /** Get an ACP job by ID */
  async getJob(jobId: string): Promise<AcpJob> {
    return this.client.get<AcpJob>(`/acp/jobs/${jobId}`);
  }

  /** List ACP jobs */
  async listJobs(options?: { agent_id?: string; phase?: string; limit?: number; offset?: number }): Promise<ListResponse<AcpJob>> {
    const data = await this.client.get<{ jobs: AcpJob[]; total: number }>('/acp/jobs', {
      agent_id: options?.agent_id,
      phase: options?.phase,
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
    });
    return { data: data.jobs, total: data.total };
  }

  /** Negotiate job terms (seller) */
  async negotiate(jobId: string, sellerAgentId: string, agreedTerms: Record<string, unknown>, agreedPriceUsdc?: number): Promise<AcpJob> {
    const body: Record<string, unknown> = { agreed_terms: agreedTerms };
    if (agreedPriceUsdc !== undefined) body.agreed_price_usdc = agreedPriceUsdc;
    return this.client.request<AcpJob>('POST', `/acp/jobs/${jobId}/negotiate`, {
      json: body,
      params: { seller_agent_id: sellerAgentId },
    });
  }

  /** Fund a job (buyer) */
  async fund(jobId: string, buyerAgentId: string): Promise<AcpJob> {
    return this.client.request<AcpJob>('POST', `/acp/jobs/${jobId}/fund`, {
      params: { buyer_agent_id: buyerAgentId },
    });
  }

  /** Deliver job results (seller) */
  async deliver(jobId: string, sellerAgentId: string, resultData: Record<string, unknown>, notes?: string): Promise<AcpJob> {
    const body: Record<string, unknown> = { result_data: resultData };
    if (notes) body.notes = notes;
    return this.client.request<AcpJob>('POST', `/acp/jobs/${jobId}/deliver`, {
      json: body,
      params: { seller_agent_id: sellerAgentId },
    });
  }

  /** Evaluate delivered work (evaluator) */
  async evaluate(jobId: string, evaluatorAgentId: string, approved: boolean, evaluationNotes?: string, rating?: number): Promise<AcpJob> {
    const body: Record<string, unknown> = { approved };
    if (evaluationNotes) body.evaluation_notes = evaluationNotes;
    if (rating !== undefined) body.rating = rating;
    return this.client.request<AcpJob>('POST', `/acp/jobs/${jobId}/evaluate`, {
      json: body,
      params: { evaluator_agent_id: evaluatorAgentId },
    });
  }

  // ── Memos ──

  /** Send a signed memo on a job */
  async sendMemo(jobId: string, senderAgentId: string, memoType: string, content: Record<string, unknown>, signature?: string): Promise<AcpMemo> {
    const body: Record<string, unknown> = { memo_type: memoType, content };
    if (signature) body.signature = signature;
    return this.client.request<AcpMemo>('POST', `/acp/jobs/${jobId}/memos`, {
      json: body,
      params: { sender_agent_id: senderAgentId },
    });
  }

  /** List memos for a job */
  async listMemos(jobId: string): Promise<ListResponse<AcpMemo>> {
    const data = await this.client.get<{ memos: AcpMemo[]; total: number }>(`/acp/jobs/${jobId}/memos`);
    return { data: data.memos, total: data.total };
  }

  // ── Resource Offerings ──

  /** Create a resource offering */
  async createOffering(params: CreateOfferingParams): Promise<ResourceOffering> {
    return this.client.post<ResourceOffering>('/acp/offerings', {
      agent_id: params.agent_id,
      name: params.name,
      description: params.description,
      endpoint_path: params.endpoint_path,
      parameters: params.parameters ?? {},
      response_schema: params.response_schema ?? {},
    });
  }

  /** List resource offerings */
  async listOfferings(options?: { agent_id?: string; limit?: number; offset?: number }): Promise<ListResponse<ResourceOffering>> {
    const data = await this.client.get<{ offerings: ResourceOffering[]; total: number }>('/acp/offerings', {
      agent_id: options?.agent_id,
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
    });
    return { data: data.offerings, total: data.total };
  }
}
